### PR TITLE
Show special BonApp menu messages for closed cafs

### LINKED ISSUE #3373: Show BonApp special message for closed cafeterias
Right now, if the menu retrieved is empty, we always return the vague message "No items found." Instead, we should see if BonApp has provided a special message, which is often more descriptive, and if so, show that instead, along with the date.

Current Behavior | Desired Behavior
-- | --
screen shot 2018-12-29 at 5 02 21 pm | screen shot 2018-12-29 at 5 23 12 pm

### PR DESCRIPTION
**Pull request by hannesmcman**
Resolves #3373.

### CODE REVIEW

**Review by drewvolz — CHANGES_REQUESTED** [review]
This reverts some logic that still needs to be taken into account.

File: source/views/menus/menu-bonapp.js
Comment (drewvolz) [inline source/views/menus/menu-bonapp.js:264]: Right, so I don't think we can outright remove this. I added it originally back in #3047 because we have to handle all ranges of cafes provided by bonapp. Sometimes there are no dayparts and no items, yet the cafeteria is open.

I'm testing this with cafe id 2 and the error that this block of code originally helped catch is back without it. So while it may cause another issue where it's at now, this still needs to be taken into account _somewhere_.

Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js:264]: Ahh I see..My latest commit should fix this issue. I tested it with cafe id 2 and it works as expected.

**Review by hawkrives — COMMENTED** [review]
I haven't fully looked through this, but I just want to make sure that this doesn't hide the toolbar when you're looking at a closed menu (as opposed to a closed cafe).

**Comment by hannesmcman** [comment]
If there are no food items, only the top portion of the toolbar will show, the part with the date in it. 

screen shot 2018-12-29 at 8 33 20 pm

**Comment by hawkrives** [comment]
That would prevent you from changing menus, then, if you wanted to check Lunch and Breakfast was empty?

**Comment by hannesmcman** [comment]
Sorry I should be more clear.. the only part of the filter toolbar that will be hidden is the part where the filters themselves can be changed. If the location offers multiple meals that toggle will still be available, which is something that is not currently possible before this PR.

https://github.com/StoDevX/AAO-React-Native/blob/31fbc658ca5cd4673139f614046e8afd5ea5e6ea/modules/food-menu/filter-menu-toolbar.js#L46-L69

File: source/views/menus/menu-bonapp.js
Comment (drewvolz) [reply source/views/menus/menu-bonapp.js:264]: Okay. I'm noticing that we are unable to parse cafe ids 0 and 3000. The cafes returns `{"cafe":[]}`. We should add a check for no cafe data...

Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js:264]: Good point! The latest commit should resolve this..

Comment (drewvolz) [reply source/views/menus/menu-bonapp.js:264]: I don’t know how many edge cases this spans to

Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js:264]: What kinds of edge cases are you envisioning?

**Comment by hannesmcman** [comment]
Jest test failures seem to be unrelated to the contents of this PR...

**Comment by hawkrives** [comment]
Re:jest; agreed. Can you reproduce locally?

**Comment by hawkrives** [comment]
Re:menu switching, ok cool. Sgtm.

**Comment by hannesmcman** [comment]
Re:jest, Yes I can reproduce locally...looking into it now, but could use some help as I did not write these tests or the code that they are testing..

**Comment by rye** [comment]
Judging by the Jest output it looks like this is related to #3363, which was merged a bit ago. Does Jest run fine on `master`?

**Comment by rye** [comment]
Oh I bet I know what it is, we had build issues last year around new year too, didn't we.

**Comment by hannesmcman** [comment]
Ah yes I remember that...not sure exactly why these tests are failing though

**Comment by rye** [comment]
Okay given I've spent ~1h looking at the tests and trying to make them green again to no avail, I think that as long as no `building-hours` or `bus-schedule`-related failures show up in your Jest output, they're safe to ignore. But obviously a build will not continue to native if that is not the case… Hmm.

**Comment by rye** [comment]
Once #3376 gets merged, I think the Jest failure should go away.

File: modules/food-menu/fancy-menu.js
Comment (rye) [inline modules/food-menu/fancy-menu.js:192]: Please add a newline separating the previous line and this line.

File: source/views/menus/menu-bonapp.js
Comment (rye) [inline source/views/menus/menu-bonapp.js:269]: New line before this line please.

File: source/views/menus/menu-bonapp.js
Comment (rye) [inline source/views/menus/menu-bonapp.js:255]: Type checking is a smell but it wasn't added here so don't mind this comment.

File: source/views/menus/menu-bonapp.js
Comment (rye) [inline source/views/menus/menu-bonapp.js:224]: What happens if `dayparts.length === 0`? Then this becomes `null` or `undefined`, I guess. Is that handled?

File: source/views/menus/menu-bonapp.js
Comment (rye) [inline source/views/menus/menu-bonapp.js:221]: What happens if `dayparts.length === 0`? Then this becomes `null` or `undefined`, I guess. Is that handled?

File: modules/food-menu/filter-menu-toolbar.js
Comment (rye) [inline modules/food-menu/filter-menu-toolbar.js]: Could this be changed to `isOpen` instead? `!isClosed` is a double negative.

**Review by rye — COMMENTED** [review]
Some nits.

File: source/views/menus/menu-bonapp.js
Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js:221]: If `dayparts.length === 0`, the ternary "short-circuits" and doesn't bother checking the second condition and the condition is evaluated as `false`. Right? @rye

Comment (rye) [reply source/views/menus/menu-bonapp.js:221]: Well for one, I don't think this logic covers the case where `dayparts[0]` doesn't have a `.length`?

In any case, here's what I get when `dayparts.length === 0`:

```console
> let ary = []
undefined
> ary.length !== 0 && ary[0].length ? ary[0] : 'default'
'default'
```

***

Ohh… precedence here means the condition on the ternary is `dayparts.length !== 0 && dayparts[0].length`, not `dayparts[0].length`. (My bad.)

You're right, the `&&` stops evaluating after the first condition becomes `false`.

My new question is: can you wrap the condition in parens? (Will prettier change it back?) It's more readable that way.

**Comment by rye** [comment]
@hannesmcman, I'll fire up a sim and give this a look-see now. Thanks for resolving those nits.

File: source/views/menus/menu-bonapp.js
Comment (drewvolz) [reply source/views/menus/menu-bonapp.js:264]: I guess this actually covers it 👍

File: source/views/menus/menu-bonapp.js
Comment (hawkrives) [reply source/views/menus/menu-bonapp.js:221]: Prettier will remove the parens.

Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js:221]: If `dayparts[0]` doesn't have a `.length`, that part of the condition will return `undefined`, which will make the whole condition evaluate as `undefined` and so the ternary will return the default schedule, as desired.

**Review by drewvolz — APPROVED** [review]

**Review by rye — APPROVED** [review]
Looks to be operating as intended to me, but please wait for additional feedback from @hawkrives and @drewvolz. BonApp's legacy API is serving us seemingly faulty information over breaks. (I'm specifically looking at Stav, which is marked as "open" on this API:)

image

As to the functionality of this, it definitely seems to be doing the right thing. The Legacy API is just inconsistent with their website, which was what confused me.

**Comment by hawkrives** [comment]
Dammit. @rye, can you check to see if there's a new API powering their website, please, or if it's just server-side HTML?

File: source/views/menus/menu-bonapp.js
Comment (hawkrives) [inline source/views/menus/menu-bonapp.js]: Array.isArray(stuff), please

**Review by hawkrives — DISMISSED** [review]
lgtm, pending comment

File: source/views/menus/menu-bonapp.js
Comment (hannesmcman) [reply source/views/menus/menu-bonapp.js]: sure

**Comment by rye** [comment]
@hawkrives, their website appears to be driven by its own HTML. I believe this because when I opened it, I didn't see any calls to anything like `legacy.cafebonappetit.com`, and everything was HTML or otherwise. No JSON.

**Comment by drewvolz** [comment]
something something the site is wordpress

something something there are network requests but maybe server-side because `Bamco.menu_items` is already populated?
```js
<script type="text/javascript">
(function() {
	Bamco = (typeof Bamco !== "undefined") ? Bamco : {};
	Bamco.api_url = {
		items: 'https://legacy.cafebonappetit.com/api/2/items?format=jsonp',
		cafes: 'https://legacy.cafebonappetit.com/api/2/cafes?format=jsonp',
		menus: 'https://legacy.cafebonappetit.com/api/2/menus?format=jsonp'
	};
	Bamco.current_cafe = {
		name: 'Stav Hall',
		id: 243				};
	Bamco.view_tier = 1;
	Bamco.menu_items = [redacted];
                        Bamco.dietary_filters = [];
    			})();
</script>
```

**Comment by rye** [comment]
@drewvolz the uh

https://legacy.cafebonappetit.com/api/2/cafes?cafe=261

That's what I was talking about tho.

So they're doing their own override thing? That'd explain it—if they don't fetch unless they're missing stuff… But the site still says Stav's closed whereas the API says it's open.

Still, it's BonApp's fault, I think. 😈 But maybe we should think about doing more than just an API request, e.g. more rigorous closure checks.

**Comment by drewvolz** [comment]
On another note, from my code snippet, why does their script say 243 is stav’s id?

**Pull request merged**
